src/main/groovy/com/wooga/gradle/PropertyLookup.groovy

```diff
@@ -195,36 +289,32 @@ class PropertyLookup {
         )
     }
 
-    Provider<String> getStringValueProvider(Project project) {
-        project.provider({
-            getStringValueProvider(project.getProviders(), project.properties, System.getenv())
-        }).flatMap({it})
-    }
-
-    Provider<Boolean> getBooleanValueProvider(Project project) {
-        project.provider ({
-            getBooleanValueProvider(project.getProviders(), project.properties, System.getenv())
-        }).flatMap({it})
-    }
-
-    Provider<RegularFile> getFileValueProvider(Project project) {
-        project.provider({
-            getFileValueProvider(project.providers, project.layout, project.properties, System.getenv())
-        }).flatMap({it})
-    }
-
+    /**
+     * @return A provider which returns a {@code Directory}
+     */
     Provider<Directory> getDirectoryValueProvider(Project project) {
-        project.provider( {
+        project.provider({
             getDirectoryValueProvider(project.providers, project.layout, project.properties, System.getenv())
-        }).flatMap({it})
+        }).flatMap({ it })
     }
 
-    static String envNameFromProperty(String extensionName, String property) {
-        "${extensionName.toUpperCase()}_${property.replaceAll(/([A-Z])/, "_\$1").toUpperCase()}"
+    /**
+     * @return A provider which returns an object of type {@code T} based on the given closure
+     */
+    public <T> Provider<T> getValueProvider(ProviderFactory factory, Map<String, ?> properties, Function<String, T> parseFunc, Map<String, ?> env = null) {
+        factory.provider({
+            def rawValue = getValue(properties, env).toString()
+            def value = parseFunc.apply(rawValue)
+            value
+        }) as Provider<T>
     }
 
-    static String convertPropertyToEnvName(String property) {
-        property.replaceAll(/([A-Z.])/, '_$1').replaceAll(/[.]/, '').toUpperCase()
+    /**
+     * @return A provider which returns an object of type {@code T} based on the given closure
+     */
+    public <T> Provider<T> getValueProvider(Project project, Function<String, T> parseFunc) {
+        project.provider({
+            getValueProvider(project.providers, project.properties, parseFunc, System.getenv())
+        }).flatMap({ it })
```

Comment:
The `public` keywords are not needed. I don't really get why we would need this. The rawValue will alwas be a `String`. So one could simply fetch the value as string and call `map` on the property. Als your code could be rewritten with the mapper types of gradle

as in something like this:
```groovy
(factory.provider({
           getValue(properties, env).toString()
}) as Provider<T>).map(parseFunc)
```

Which means that parseFunc/map function would be of Type `org.gradle.api.Transformer<T, String>`. But again I don't see the bigger benefit of having this API internal if it is just one offical call.

Reply:
I actually had to add the public keywords in these generics for it to work; the IDE kept showing errors otherwise.

![image](https://user-images.githubusercontent.com/10147802/156343046-f687c890-429a-40c8-a027-fe26bf3de3ab.png)

From an API perspective, it's much nicer to just pass the closure and have the type deduced then having to do the map invocation yourself. As seen here:

![image](https://user-images.githubusercontent.com/10147802/156342953-67492bb6-3e96-4be5-84ff-e3094da0e2b7.png)

The rawValue is not guaranteed to be a string if it ended up coming from the default value of the property, which could be whatever was set during the assignment.

Reply:
But that is the benefit of functional programming. I mean yes we can have this API but why adding it when the exported type already provides one. I would make sense if the class would allow to be extended with concrete functions for custom mappings. And in you example it would maybe make more sense to implement this for a given enum class. But I don't know if this is actually possible.

`public <T extends Enum> Provider<T> getValueProvider(Project project, T enumClass)`

But I think we have issues there making this actually generic.
As far as I see it it will mainly be only sugar. But we can put it in. But I would argue to change the signature to support the official gradle API instead of calling the mapper in the closure.

Reply:
I got various warnings when trying to write the function calling map at the end.
In addition, wouldn't that have us creating two providers rather than just one if we call the parse function within the initial closure?

Reply:
If this gets warning and other things I would just leave it out. Sure you would loose a function call. But map does also a null check. Means if the result of the first operation yields null then the whole provider chain yields null. You don't have a null check and also it is not garanteet for the closure if it has to do a null check. I see the reason to having this but just for a few characters saved ... But overall it doesn't hurt to have it if you feel it will make things easier.

Reply:
I will merge it now and evaluate the internals of the function; whether we would like to do the transformation.